File: practice_java_program/notpad/LongestKDistinct.java
//Longest Substring with At Most K Distinct Characters
import java.util.*;
public class LongestKDistinct{
    public static void main(String[] args) {
        String s = "eceba";
        int k = 2;
        HashMap<Character, Integer> map = new HashMap<>();

        int left = 0, maxLen = 0;

        for (int right = 0; right < s.length(); right++) {

            char ch = s.charAt(right);
            map.put(ch, map.getOrDefault(ch, 0) + 1);

            while (map.size() > k) {

                char leftChar = s.charAt(left);
                map.put(leftChar, map.get(leftChar) - 1);

                if (map.get(leftChar) == 0) {
                    map.remove(leftChar);
                }
                left++;
            }
            maxLen = Math.max(maxLen, right - left + 1);
        }
        System.out.println("Longest Length: " + maxLen);
    }
}

File: python_oriented_interview/notpad/LongestKDistinct.py
s = "eceba"
k = 2
map_dict = {}

left = 0
maxLen = 0

for right in range(len(s)):
    ch = s[right]
    map_dict[ch] = map_dict.get(ch, 0) + 1

    while len(map_dict) > k:
        leftChar = s[left]
        map_dict[leftChar] -= 1
        if map_dict[leftChar] == 0:
            del map_dict[leftChar]
        left += 1

    maxLen = max(maxLen, right - left + 1)

print("Longest Length:", maxLen)
